### PR TITLE
IBD: do not check PoW (Yespower) during Download headers

### LINKED ISSUE #45: too much traffic during IBD
image

image

actual blockchainsize is `1.4 GB`, but during IBD it uses too much traffic (around `40 GB`). so mobile users data limit ran out too fast.

```
Fais Sheikh, [09.01.20 12:48]
How big sugar wallet size ?

Fais Sheikh, [09.01.20 12:48]
100 gb?

Fais Sheikh, [09.01.20 12:49]
I kept it last night to sync only 33%

Fais Sheikh, [09.01.20 12:49]
It’s 38 gb was download

Volbil, [09.01.20 12:50]
[In reply to Fais Sheikh]
0_o

Volbil, [09.01.20 12:50]
Are you sure about 38 gb?

Volbil, [09.01.20 12:50]
Not mb?

Volbil, [09.01.20 12:51]
[ Photo ]

Fais Sheikh, [09.01.20 13:06]
I kept for sync idk what went wrong

Fais Sheikh, [09.01.20 13:07]
[ Photo ]

Fais Sheikh, [09.01.20 13:07]
33% only

Fais Sheikh, [09.01.20 13:07]
I closed it cuz I had only 44gb data left lool

Fais Sheikh, [09.01.20 13:07]
Fkup

Fais Sheikh, [09.01.20 13:09]
Idk how this happened damn

Fais Sheikh, [09.01.20 13:09]
I had nothing else to sync or download

Fais Sheikh, [09.01.20 13:09]
All data eaten

Volbil, [09.01.20 13:13]
@cryptozeny we should investigate it

Volbil, [09.01.20 13:13]
@fireboy299 can you check what is inside your .sugarchain folder?

Fais Sheikh, [09.01.20 13:14]
at my wallet its only 1.6gb

Fais Sheikh, [09.01.20 13:14]
but my hdd says space went down

Fais Sheikh, [09.01.20 13:14]
not sure what file was getting donwloaded

Fais Sheikh, [09.01.20 13:15]
[ Photo ]

Volbil, [09.01.20 13:15]
[In reply to Fais Sheikh]
I think this is your network traffic

Fais Sheikh, [09.01.20 13:16]
back to 1.2?

Volbil, [09.01.20 13:16]
Not actual disk space

Fais Sheikh, [09.01.20 13:16]
NOPE

Fais Sheikh, [09.01.20 13:16]
i kept nothing else to sync

Fais Sheikh, [09.01.20 13:16]
single lap

Fais Sheikh, [09.01.20 13:16]
singlr user

Fais Sheikh, [09.01.20 13:17]
its my protable net

Fais Sheikh, [09.01.20 13:17]
so i use for lap only

Fais Sheikh, [09.01.20 13:17]
limited gb

cryptozeny, [09.01.20 13:18]
[In reply to Fais Sheikh]
@Volbil yeah i noticed duplicated download. it download multiple times.

Fais Sheikh, [09.01.20 13:19]
where is this file

Fais Sheikh, [09.01.20 13:19]
i need to remove

cryptozeny, [09.01.20 13:20]
[In reply to Fais Sheikh]
total blockchain size is currently around 1.5 gb. another is not stored

cryptozeny, [09.01.20 13:20]
[In reply to Fais Sheikh]
you downloaded, but its just check block verifying

cryptozeny, [09.01.20 13:21]
not stored in your disk

Fais Sheikh, [09.01.20 13:21]
ok

Fais Sheikh, [09.01.20 13:21]
what to do

Fais Sheikh, [09.01.20 13:21]
i dont need other stuff

cryptozeny, [09.01.20 13:22]
actually Yumekawa 4x times verifying all blocks since genesis.

Fais Sheikh, [09.01.20 13:22]
i just want to sync

Fais Sheikh, [09.01.20 13:22]
i will see where is this file at pc

cryptozeny, [09.01.20 13:22]
[In reply to Fais Sheikh]
if you have limited data transfer, just copy from already synced.

Fais Sheikh, [09.01.20 13:23]
where can i get it

cryptozeny, [09.01.20 13:24]
[In reply to Fais Sheikh]
Windows: C:\Users\YourUserName\Appdata\Roaming\Sugarchain\ Linux: ~/.sugarchain/ OSX: ~/Library/Application Support/Sugarchain/

cryptozeny, [09.01.20 13:24]
blocks and chainstate folder stotes all blockchain data

cryptozeny, [09.01.20 13:25]
but do noy delete wallet.dat its your actuall wallet

cryptozeny, [09.01.20 13:26]
you found an over traffic than current blockchain size. this will fixed in next release.

cryptozeny, [09.01.20 13:27]
[In reply to Volbil]
right. its network traffic

cryptozeny, [09.01.20 13:30]
[In reply to Fais Sheikh]
actually 1.4 gb for now. otherwise its jusy network traffic.
https://twitter.com/sugarchain_dev/status/1212483415410274307?s=19

cryptozeny, [09.01.20 13:31]
your traffic will be bigger than this to verify many times.

Fais Sheikh, [09.01.20 14:13]
Ok noted

Fais Sheikh, [09.01.20 14:13]
Sync where to download
```

### PR DESCRIPTION
**Pull request by decryp2kanon**
# Background
https://github.com/sugarchain-project/sugarchain/pull/119

# TL;DR
- CURRENT: 
  * when downloading headers, check header (PoW: Yespower)
  * bottle neck at downloading headers
- PR:
  * skip check header (PoW: Yespower), only in downloading headers
  * bottle neck gone! :tada: 

# Safe to skip?
`CheckProofOfWork()` is already multiple checking in several place. i believe that even if our node is attacked, it will defense well.

-----

# RESULT
- :heavy_check_mark: total IBD time `++23%` faster
image

- :heavy_check_mark: data traffic `--60%` reduced
  * before
    image

  * after  
    image

- :heavy_check_mark: less CPU usage around `--5%` when IBD

-----

# What we are skipping?
skipped check `CheckProofOfWork()` at `pow.cpp` **when Downloading headers**
```cpp
bool CheckProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params& params)
{
    bool fNegative;
    bool fOverflow;
    arith_uint256 bnTarget;

    bnTarget.SetCompact(nBits, &fNegative, &fOverflow);

    // Check range
    if (fNegative || bnTarget == 0 || fOverflow || bnTarget > UintToArith256(params.powLimit))
        return false;

    // Check proof of work matches claimed amount
    if (UintToArith256(hash) > bnTarget)
        return false;

    return true;
}
```

# How to skip and when?
do not PoW check when **Downloading headers of IBD (include `fImporting` and `fReindex`)**. if not, we do not skip.
```cpp
if ( IsInitialBlockDownload() == false )
```

# Safe enough?
i believe so. its already multiple checked in `CheckBlock()` and this can be found in 4 places:

<details><summary>CLICK ME</summary>
<p>

- 1. `AcceptBlock()`
```cpp
    if (!CheckBlock(block, state, chainparams.GetConsensus()) ||
        !ContextualCheckBlock(block, state, chainparams.GetConsensus(), pindex->pprev)) {
        if (state.IsInvalid() && !state.CorruptionPossible()) {
            pindex->nStatus |= BLOCK_FAILED_VALID;
            setDirtyBlockIndex.insert(pindex);
        }
        return error("%s: %s", __func__, FormatStateMessage(state));
    }
```
- 2. `TestBlockValidity()`
```cpp
    // NOTE: CheckBlockHeader is called by CheckBlock
    if (!ContextualCheckBlockHeader(block, state, chainparams, pindexPrev, GetAdjustedTime()))
        return error("%s: Consensus::ContextualCheckBlockHeader: %s", __func__, FormatStateMessage(state));
    if (!CheckBlock(block, state, chainparams.GetConsensus(), fCheckPOW, fCheckMerkleRoot))
        return error("%s: Consensus::CheckBlock: %s", __func__, FormatStateMessage(state));
```

- 3. `VerifyDB()`
```cpp
        // check level 1: verify block validity
        if (nCheckLevel >= 1 && !CheckBlock(block, state, chainparams.GetConsensus()))
            return error("%s: *** found bad block at %d, hash=%s (%s)\n", __func__,
                         pindex->nHeight, pindex->GetBlockHash().ToString(), FormatStateMessage(state));
```

- 4. `ConnectBlock()`
```cpp
    // Check it again in case a previous version let a bad block in
    // NOTE: We don't currently (re-)invoke ContextualCheckBlock() or
    // ContextualCheckBlockHeader() here. This means that if we add a new
    // consensus rule that is enforced in one of those two functions, then we
    // may have let in a block that violates the rule prior to updating the
    // software, and we would NOT be enforcing the rule here. Fully solving
    // upgrade from one software version to the next after a consensus rule
    // change is potentially tricky and issue-specific (see RewindBlockIndex()
    // for one general approach that was used for BIP 141 deployment).
    // Also, currently the rule against blocks more than 2 hours in the future
    // is enforced in ContextualCheckBlockHeader(); we wouldn't want to
    // re-enforce that rule here (at least until we make it impossible for
    // GetAdjustedTime() to go backward).
    if (!CheckBlock(block, state, chainparams.GetConsensus(), !fJustCheck, !fJustCheck))
        return error("%s: Consensus::CheckBlock: %s", __func__, FormatStateMessage(state));
```

</p>
</details>

# Screenshot during IBD
- 1. This means PoW check during IBD is not actually skipped, but still its checking.
image

### CODE REVIEW

**Comment by decryp2kanon** [comment]
image

`IBD=1` this means its still checking PoW during IBD. CBH for `CheckBlockHeader()` that includes `CheckProofOfWork()`

**Comment by decryp2kanon** [comment]
when do PoW check?
1. IBD to first user :heavy_check_mark: 
2. reindex-chainstate (fReindex) :heavy_check_mark: 
3. reindex (fReindex) :heavy_check_mark: 
4. fImporting :question: 

image

**Comment by decryp2kanon** [comment]
- black: moonlight (current)
- red: IBD
- green: `-reindex-chainstate`
- blue: soon

image

**Comment by decryp2kanon** [comment]
image

**Comment by akshaynexus** [comment]
Looks like a good change to me,i had alot of delay and cpu usage when syncing

**Comment by decryp2kanon** [comment]
- moonlight
- skip: OK
- reindex: OK
- reindex-chainstate: OK

image

`12.5%` faster than moonlight

**Comment by decryp2kanon** [comment]
Tested ACK 032184c4e70cb53b7c817bc2e65dcc6b3541f9a2

**Comment by decryp2kanon** [comment]
IBD around `10~30%` faster, however it does not skip checking PoW, but only remove duplicated at downloading stage

**Comment by decryp2kanon** [comment]
reference about 
```cpp
if (!x() && !y())
```
https://stackoverflow.com/questions/8300519/c-if-statement-with-which-function-will-execute-first

File: src/validation.cpp [hunk 1]
```diff
@@ -3334,7 +3341,9 @@ bool CChainState::AcceptBlockHeader(const CBlockHeader& block, CValidationState&
             return true;
         }
 
-        if (!CheckBlockHeader(block, state, chainparams.GetConsensus()))
+        // FIXME.SUGAR // check PoW: SKIPPED during downloading headers (IBD)
+        // IBD: do not check PoW (Yespower) during Download headers for performance reason
+        if (!IsInitialBlockDownload() && !CheckBlockHeader(block, state, chainparams.GetConsensus()))
```
Comment (decryp2kanon) [inline src/validation.cpp:3346]:
```cpp
if ( !CheckBlockHeader(block, state, chainparams.GetConsensus()) ( !IsInitialBlockDownload() ) )
```
@volbil

Comment (decryp2kanon) [inline src/validation.cpp:3347]:
```cpp
// Case #2 - We run CheckBlockHeader function, but if it's IBD - fCheckPOW set to false and CheckProofOfWork is ignored
bool idb = IsInitialBlockDownload()
if (!CheckBlockHeader(block, state, chainparams.GetConsensus(), !idb))
    return error("%s: Consensus::CheckBlockHeader: %s, %s", __func__, hash.ToString(), FormatStateMessage(state));
```

https://gist.github.com/itwysgsl/982d065a2fea42d9e0238932eab265d4

**Comment by decryp2kanon** [comment]
https://github.com/sugarchain-project/sugarchain/blob/032184c4e70cb53b7c817bc2e65dcc6b3541f9a2/src/validation.cpp#L3067-L3070

this comment should be changed

**Review by volbil — APPROVED** [review]
Looking good to me 👍

**Comment by Nugetzrul3** [comment]
Currently running new version. Noticeably quicker startup 👍

**Comment by decryp2kanon** [comment]
https://gist.github.com/itwysgsl/982d065a2fea42d9e0238932eab265d4

nice suggestion. thanks! volbil

File: src/validation.cpp (hunk 1 above)
Comment (decryp2kanon) [reply src/validation.cpp:3347]: both seems exact same result.
